walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,85 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
+  Collection<TimeRange> possibleMeetingTime = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // find the list of attendees and required duration
+    Collection<String> reqAttendees = request.getAttendees();
+    int reqDuration = (int)request.getDuration();
+
+    // corner case: no attendees in the request
+    if (reqAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // corner case: requested duration more than the day duration
+    if (reqDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // check if each event's attendees clash with the request attendees sorted by starting time
+    List<TimeRange> eventTimes = events.stream()
+                                      .filter(event -> isAffectedEvent(event.getAttendees(), reqAttendees))
+                                      .map(event -> event.getWhen())
+                                      .sorted(TimeRange.ORDER_BY_START)
+                                      .collect(Collectors.toList());
+
+    // no event affecting required attendees
+    if (eventTimes.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    int possibleStartTime = TimeRange.START_OF_DAY;
+
+    // check first event starting time
```

Comment:
In general, I'd avoid comments that just state exactly what is being done, since it's self evident from the code. (Except sometimes for class or method javadoc comments to give an overview). However, it would be helpful to have a comment in this case like maybe `// Consider the time window between the start of the day and the first event.`